src/canvas.js

```diff
@@ -6,18 +6,29 @@ export default class Canvas {
    * Initialise.
    * @param {string} [canvasId]
    *   An optional CSS ID pointing to a canvas to override the default.
+   * @param {HTMLCanvasElement} [canvasId]
+   *   An optional HTMLCanvasElement to override the default.
    */
-  constructor(canvasId) {
+  constructor(canvasId, canvasElement) {
     const defaultCanvasId = 'confetti-canvas';
-    const customCanvas = document.getElementById(canvasId);
 
-    this.isDefault = canvasId === null || typeof canvasId === 'undefined';
+    if (canvasElement) {
```

Comment:
This constructor is getting a little confusing with the nested if statements, perhaps something like this would be a little more readable?

```js
constructor(canvasId, canvasElement) {
  const defaultCanvasId = 'confetti-canvas';
  let canvas = canvasElement;

  if (canvasElement && !(canvasElement instanceof HTMLCanvasElement)) {
    throw new Error('Element is not a valid HTMLCanvasElement');
  }

  if (!canvas) {
    canvas = document.getElementById(canvasId);
  }

  if (canvasId && !canvas) {
    throw new Error(`No element found with ID "${canvasId}"`);
  }

  this.isDefault = !(canvasId || canvasElement);

  this.canvas = canvas || Canvas.createDefaultCanvas(defaultCanvasId);

  this.ctx = this.canvas.getContext('2d');
}
```
Although, see comment below, might be able to just leave it as it is.

Reply:
@alexandermendes should check to ensure canvas element retrieved is of type `HTMLCanvasElement` or not, so constructor should be something like this?
```js
  constructor(canvasId, canvasElement) {
    const defaultCanvasId = 'confetti-canvas';
    let canvas = canvasElement;

    if (canvasElement && !(canvasElement instanceof HTMLCanvasElement)) {
      throw new Error('Element is not a valid HTMLCanvasElement');
    }

    if (canvasId && !canvas) {
      canvas = document.getElementById(canvasId);
    }

    if (canvasId && !canvas) {
      throw new Error(`No element found with ID "${canvasId}"`);
    }

    if (!(canvas instanceof HTMLCanvasElement)) {
      throw new Error(`Element with ID "${canvasId}" is not a valid HTMLCanvasElement`);
    }

    this.isDefault = !(canvasId || canvasElement);

    this.canvas = canvas || Canvas.createDefaultCanvas(defaultCanvasId);

    this.ctx = this.canvas.getContext('2d');
  }
```

Reply:
Yeah fair point, that could be worth an extra check. Although at that point `canvas` may still be undefined, if we are reverting to the default. We could merge your extra check into the one before it, so:

```js
constructor(canvasId, canvasElement) {
    const defaultCanvasId = 'confetti-canvas';
    let canvas = canvasElement;

    if (canvasElement && !(canvasElement instanceof HTMLCanvasElement)) {
      throw new Error('Element is not a valid HTMLCanvasElement');
    }

    if (canvasId && !canvas) {
      canvas = document.getElementById(canvasId);
    }

    if (canvasId && !(canvas instanceof HTMLCanvasElement) {
      throw new Error(`Element with ID "${canvasId}" is not a valid HTMLCanvasElement`);
    }

    this.isDefault = !(canvasId || canvasElement);

    this.canvas = canvas || Canvas.createDefaultCanvas(defaultCanvasId);

    this.ctx = this.canvas.getContext('2d');
  }
```